### PR TITLE
[SPARK-11691][SQL] Support setting hadoop compression codecs in DataFrameWriter#option

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?

This pr is to support hadoop compression codecs when saving DataFrame to disk.
This is rework from #9657 because it gets stale.

closes #9657

### CODE REVIEW

**Comment by maropu** [comment]
@rxin Could you check this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelation.scala [hunk 1]
```diff
@@ -58,7 +59,8 @@ import org.apache.spark.util.Utils
 private[sql] case class InsertIntoHadoopFsRelation(
     @transient relation: HadoopFsRelation,
     @transient query: LogicalPlan,
-    mode: SaveMode)
+    mode: SaveMode,
+    codec: Option[Class[_ <: CompressionCodec]] = None)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelation.scala]: codec -> compressionCodec to make it more clear

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 2]
```diff
@@ -239,8 +246,9 @@ private[sql] abstract class BaseWriterContainer(
 private[sql] class DefaultWriterContainer(
     relation: HadoopFsRelation,
     job: Job,
-    isAppend: Boolean)
-  extends BaseWriterContainer(relation, job, isAppend) {
+    isAppend: Boolean,
+    codec: Option[Class[_ <: CompressionCodec]])
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: same thing here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 3]
```diff
@@ -308,8 +316,9 @@ private[sql] class DynamicPartitionWriterContainer(
     inputSchema: Seq[Attribute],
     defaultPartitionName: String,
     maxOpenFiles: Int,
-    isAppend: Boolean)
-  extends BaseWriterContainer(relation, job, isAppend) {
+    isAppend: Boolean,
+    codec: Option[Class[_ <: CompressionCodec]])
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: same thing here

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala [hunk 4]
```diff
@@ -57,6 +57,13 @@ class TextSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("compression") {
+    val tempDirPath = Utils.createTempDir().getAbsolutePath
+    val df = sqlContext.read.text(testFile)
+    df.write.option("compressionCodec", "GzipCodec").mode(SaveMode.Overwrite).text(tempDirPath)
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: what are the mappings provided here? I personally find "GzipCodec" pretty annoying to use. Maybe we can create short keys for these, similar to how we handle compression elsewhere.

**Comment by rxin** [comment]
@maropu can you add "closes #9657" to your pull request description?

**Comment by maropu** [comment]
okay, I'm fixing now.

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
ISTM the fail is not related to this...
```
[info] HiveCompatibilitySuite:
[info] - lateral_view *** FAILED *** (933 milliseconds)
[info]   Failed to execute query using catalyst:
[info]   Error: Job aborted due to stage failure: Task 0 in stage 8347.0 failed 1 times, most recent failure: Lost task 0.0 in stage 8347.0 (TID 120049, localhost): java.lang.ClassCastException: org.apache.spark.sql.catalyst.expressions.JoinedRow cannot be cast to org.apache.spark.sql.catalyst.expressions.UnsafeRow
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala (hunk 4 above)
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: @rxin how about this fix?

**Comment by rxin** [comment]
Take a look at the end of ParquetRelation. It would be great if we can consolidate the two and put in a single place where we specify short compression codec names.

**Comment by rxin** [comment]
Yea I reverted a patch that broke the test.

**Comment by maropu** [comment]
Okay, I'll try.

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
@rxin okay, ready to review again.

**Comment by HyukjinKwon** [comment]
@maropu @zjffdu @rxin I apologize that I carelessly open the same issue and submitted a PR. This is fixed in https://github.com/apache/spark/pull/11384.

**Comment by maropu** [comment]
@rxin @HyukjinKwon Is it okay to apply my changes based on the @HyukjinKwon commit?

**Comment by rxin** [comment]
Yes - please.

**Comment by maropu** [comment]
@HyukjinKwon Is it okay to include document descriptions for codec options in this pr?

**Comment by maropu** [comment]
Close this pr, and move to a new pr.

**Pull request closed without merging**